### ISSUE TITLE
PEP8 and line length

### ISSUE THREAD

**Issue by XorUnison** [ManimCommunity/manim#174]
There's the intent and a PR #172 flying around to use an autopep8 library to automatically format the code into PEP8 guidelines. I want to have a discussion about two things relating to this PEP8 implementation, both relating to line length specifically.

1. By default the PEP8 line length is 79 for code and 72 for docstrings/comments. The question is, do we actually want it that restrictively short? According to official PEP8 guidelines extending code length up to 99 is just fine.
For more context, this small 79 characters limit means it'll even try to split up lines like this one in main.py
```python
            spec = importlib.util.spec_from_file_location(module_name, file_name)
```
```python
            spec = importlib.util.spec_from_file_location(
                module_name, file_name)
```
 
 
2. Whatever the line length is, there is still an issue with autopep8 we need to be aware of when contributing. Example:
```python
        logger.warning("FadeOutAndShiftDown is deprecated and will eventually disappear. Please use FadeOutAndShift(<mobject>, direction=DOWN, <other_args>) instead.")
```
Adding long strings like that in "String" format only leads autopep8 to throw the whole string down like:
```python
        logger.warning(
            "FadeOutAndShiftDown is deprecated and will eventually disappear. Please use FadeOutAndShift(<mobject>, direction=DOWN, <other_args>) instead.")
```
Which doesn't really get us anywhere. The line is still far too long, and isn't really more readable either.
Similarly it also keeps formats like:
```python
            raise Exception(
                "ChangingDecimal can only take "
                "in a DecimalNumber"
            )
```
Which I think doesn't look are read too great either.
In order to actually keep with the line lengths we'll have to either go with multiple "String" lines, or what I think looks best, just directly use """String""" for stuff that's otherwise too long, which I hope autopep8 can work with as well.
```python
                raise ValueError("""ArcBetweenPoints called with a radius that is
                            smaller than half the distance between the points.""")
```
(Although I suppose the indent of the second line would have to be different.)

As for 1, I'm fairly strongly opposed to having the code line length at a measly 79, and would instead opt for the far more reasonable 99. Despite official PEP8 specifications wanting docstrings/comments at 72 explicitly, I see no reason not to raise their character limit as well, though I'm not as partial there.

**Comment by leotrs** [comment]
1. I vote for 79 or 99. I'm strongly opposed to anything longer than that.

2. If you have a very_long_string that doesn't fit within the line limit, the usual fix is to break it up, as in your second code chunk. I know you don't like this (more on this later), but this is semi-official fix. The fix is _not_ to change the whole project's code formatting guidelines just because you (or even a few of us) don't like it.

3. > I see no reason not to raise their character limit as well, though I'm not as partial there.

    The reason for this magic number 72 is that docstrings are meant to be formatted so that they look well on a terminal (usually 80 characters wide), with possibly some indentation at the beginning (1 tab is usually 8 spaces). I, for one, do most of my development on a terminal, and I very much appreciate it when software follows this particular guideline.

4. When it comes to PEP8, everyone has the one thing they disagree with. You dislike the split string format, which I think is fine. I dislike feature X, which you may be happy with. No one is going to be 100% happy with a strict code format. However, the benefits of choosing a code format far (far, FAR) outweigh the individual nuisances it causes.

**Comment by XorUnison** [comment]
> **EDIT**: I think this might have sound a bit crass. Please be sure that I'm writing with the best intentions <3

Why does this sound so familiar, lol. But don't worry, I know it's well-meant.
Anyway:
> You dislike the split string format, which I think is fine.

Maybe to clarify on this point, I prefer the """String String""" format over the "String " "String" format. But I wouldn't even go so far as to say I actively dislike the latter as it's really just a fairly small prefer. I've got no grime about adopting this format either.

I'm rather worrying about how practical it is, because I think there's at least a chance that autopep8 can handle splitting the """String String""" format properly, whereas it's absolutely clear it can't handle anything with a single ", or at least not vanilla the way paradox used it. But I don't know. If autopep8 fails on either format all the same, then it's also all the same to me.

**Comment by PgBiel** [comment]
Imo the line limit should be 120 in order to account for comments
We should limit our code to 100, but allow ourselves to go past 120 if comments require so.

**Comment by XorUnison** [comment]
Still not partial in regards to comment length but:
> ...allow ourselves to go past 120 if comments require so.
When would comments require so or would be off better? Example?

**Comment by PgBiel** [comment]
it's highly situational, but sometimes splitting comments along lines is necessary, so if we can reduce the amount of lines that are needed on those situations... also for docstrings and stuff although those have a separate rule I think so that doesn't apply

well, no matter our choice, we should try to keep our code as short as possible of course
I'm fine with both 100 and 120

**Comment by leotrs** [comment]
Can we have our code formatter have a "soft limit" of, say 99? So we get a warning when a line is longer than that, but it allows us to do it?

**Comment by PgBiel** [comment]
not sure; if it is possible to set a "hard limit" as well, then that could be good

**Comment by XorUnison** [comment]
@safinsingh was the one who set up and ran autopep8, so he should know about how the limits can be set.

**Comment by safinsingh** [comment]
I don't believe we can set soft/hard limits, but just enforce a max line length with the `--max-line-length` argument.

**Comment by Aathish04** [comment]
Since we've agreed that we should have a soft limit on line length of 99 chars and also agreed that we're using Black to autoformat our code, can this be closed?

**Comment by eulertour** [comment]
99 is actually just over halfway across my terminal by default, which would make splitting annoying 😕

Black uses 88 by default, are there any problems with that?

**Comment by Aathish04** [comment]
None whatsoever for me. Is it a hard limit or a soft limit with a warning?

**Comment by eulertour** [comment]
I don't think black has a concept of a soft limit. It's auto-formatting your code so it's not clear how a soft limit would even be used.

But from its documentation it is possible for formatted code to exceed the line limit sometimes.

**Comment by leotrs** [comment]
I'd rather have 88 than 99. @PgBiel ?

**Comment by PgBiel** [comment]
That doesn't sound like too much of a change from pep-8, either 79 or 99 here

I personally prefer 99, less than that forces you to compress more... which isnt necessarily good for readability

**Comment by eulertour** [comment]
Why 79 or 99 without any in between @PgBiel?

**Comment by PgBiel** [comment]
Just extending the limit a little bit is not enough of a reason, at least in my opinion, to break standards as established as PEP-8. As I see it, we should either stick to it, or surpass it with a (relatively) "significant" difference

**Comment by eulertour** [comment]
Well I actually don't mind sticking to it other than that I think the consensus in 2020 is that 79 is too small. I just don't want to go up to 99 since it would make screen splitting harder.

You don't have to think of it as unnecessary unless we break the standard significantly, you could also think of it as breaking standard as much as we want without causing ourselves any more problems.

**Comment by XorUnison** [comment]
If I get that right then 79-99 is just the PEP 8 range, but it doesn't have to explicitly either. And though I would take 88 over 79 any time of the day, I would still heavily prefer 99. Because:
> Well I actually don't mind sticking to it other than that I think the consensus in 2020 is that 79 is too small.

I think so too. 79 is simply awfully short and does more damage to readability than it does good. Although 88 would be enough to fix the lines of code that bothered at least me personally the most, I also think that 99 is a decent sweet spot.


> 99 is actually just over halfway across my terminal by default, which would make splitting annoying 😕

Isn't there a way to fix that via font/terminal width? I'm kinda curious what the issue exactly is so I'd appreciate a screenshot or something.

**Comment by eulertour** [comment]
This is my fullscreen terminal, with a column at 88 chars.
![printscreen](https://user-images.githubusercontent.com/43117506/87388551-e7351b80-c559-11ea-943a-79b1cbfc48ef.png)

Changing it to 99 chars pushes the bar off the right of the splits, meaning I'd have to constantly scroll to see the full lines.

Since it's already fullscreen I can't improve the situation with the terminal size (shrinking the gaps still doesn't help), and although I could make the font smaller I'd have to change it system wide since I don't use an IDE. The fact that black occasionally allows lines to surpass the specified limit will only make it that much worse.

It's not terrible but given how much time I spend working on this I'd rather not deal with the annoyance.

**Comment by PgBiel** [comment]
That's unfortunate, but I don't think this issue should be a moving factor; also consider that most lines won't actually reach 99. It's for _the greater good_

**Comment by leotrs** [comment]
My editor screen splits at 72 characters. It sucks, but I can't bring myself to make the font smaller because of reasons.

Surprisingly, a vast majority of lines don't ever reach 72 chars, so I'm with @PgBiel on this last point. And for those that do, I can usually tell what is going on at the very end so I don't *really* need to see the whole line at once.

Also, I'll just add to the conversation that having a line length limit on the shorter side actually forces you to choose better variable names and write more succinct code.

OTOH, note that adding 8 characters to the usual limit of 80 amounts to adding a whole _two levels of indentation_. If on top of that you get in the habit of writing succinct code and formatting parens and multiple lines correctly, 88 seems like enough to me.
So that's why I'm voting for 88.

**Comment by PgBiel** [comment]
> Also, I'll just add to the conversation that having a line length limit on the shorter side actually forces you to choose better variable names and write more succinct code.

But you see, the problem isn't really variable names and whatsoever, more often than not it can be an actual requirement. As I said before, and I repeat, comments will benefit the most from this, given that we will have more space to write them without having to split over lines, place comments above or below (to which there is no clear standard, and can get messy depending on the length) etc. As I said, most lines won't reach this limit, but there _are_ cases where we need it, so we should just allow them. It doesn't mean they will occur often, it just means we are letting several cases work out nicely, which will come in handy.

> OTOH, note that adding 8 characters to the usual limit of 80 amounts to adding enough space for a whole two levels of indentation.

That's not quite a lot, considering that, if we're thinking of indentation, this means we'd actually need even _more_ line length in order to accomodate more line splits due to a short limit, which can be quite contradictory. With a larger line length limit, we actually _reduce_ the amount of indentation necessary, and not due to messy code - for which other standards are supposed to solve - but mostly because it lets us organize things better if we judge it necessary.

**Comment by leotrs** [comment]
> As I said before, and I repeat, comments will benefit the most from this, given that we will have more space to write them without having to split over lines, place comments above or below (to which there is no clear standard, and can get messy depending on the length) etc.

You are right, I wasn't thinking about comments when I said that. The reason I wasn't thinking about comments is that I usually put comments above the line they are commenting, so this is never an issue for me. As you say yourself, there is no clear standard about that, so now I'm confused about your argument here. Why should we increase line length so we can avoid putting comments above, when there's no standard about that?

> As I said, most lines won't reach this limit, but there are cases where we need it, so we should just allow them.

I said the same thing, so not sure why you are throwing this back at me.

> It doesn't mean they will occur often, it just means we are letting several cases work out nicely, which will come in handy.

And this is why I suggested we inquire into the possibility of having soft/hard limits. It seems like this is not possible (at least within black), so I suggest we add a line in the contribution instructions document that asks potential contributors (as well as all devs) to try to keep lines under 80 chars as much as possible, and to break that soft limit at their discretion, especially when it comes to inline comments. Black can be then be set to 88 or 99 or whatever we decide in the end.

> That's not quite a lot, considering that, if we're thinking of indentation, this means we'd actually need even more line length in order to accomodate more line splits due to a short limit, which can be quite contradictory. 

I didn't mean to say that adding 8 characters would allow you to write lines that are the same length but start at two more levels of indentation. I'm sorry for being unclear. What I meant to say is that adding 8 characters allows you to write lines that are two-indentation-levels-wider but starting at the _same_ level of indentation. 

> With a larger line length limit, we actually reduce the amount of indentation necessary, 

I don't see how this is the case. In python, indentation means nested blocks of code. If my code needs three levels of indentation, it doesn't matter if my editor shows 80 or 120 characters wide.

> and not due to messy code - for which other standards are supposed to solve - but mostly because it lets us organize things better if we judge it necessary.

Lines wider than 80 chars constitute messy code to me. And that number is different for everyone. I think many of us in this thread keep speaking as if the width of the lines was something that we can objectively agree upon, as if one length is right and all others are wrong. This is not the case. Line length is ultimately a personal preference.

----------

We've been at this for a bit, and it's becoming clear we won't all be on the same page. We usually try to reach a consensus that is a close to unanimous as possible, but in this case I suggest we make a poll among all current @ManimCommunity/core devs. Each person gets one vote. The option that gets the most votes is implemented. Possible votes are 79, 88, 99, 120, or whatever other number of characters you think is appropriate for line length for manim. What do you all think? Should we make this poll?

**Comment by XorUnison** [comment]
> Lines wider than 80 chars constitute messy code to me. And that number is different for everyone.

But... can you really call this:
```python
            spec = importlib.util.spec_from_file_location(module_name, file_name)
```
messier than this:
```python
            spec = importlib.util.spec_from_file_location(
                module_name, file_name)
```
?

> We've been at this for a bit, and it's becoming clear we won't all be on the same page. We usually try to reach a consensus that is a close to unanimous as possible, but in this case I suggest we make a poll among all current @ManimCommunity/core devs. Each person gets one vote. The option that gets the most votes is implemented. Possible votes are 79, 88, 99, 120, or whatever other number of characters you think is appropriate for line length for manim. What do you all think? Should we make this poll?

As long as it's a multiple choice or even better a ranked poll.
Singular choice is as terrible for this kind of stuff as it is in democracy in general and wouldn't get us anywhere.

**Comment by eulertour** [comment]
As long as we don't agree on a choice we should just start with strict PEP-8 compliance and change from there as desired. This decision has already been made, and without agreement on how to change it we can just go with the standard.

To respond to some of the earlier points:
> That's unfortunate, but I don't think this issue should be a moving factor; also consider that most lines won't actually reach 99. It's for the _greater good_

With any decision like this we should focus on the arguments for or against each point. This viewpoint is fine, but saying it without any justification doesn't hold much weight. I'd argue that the only purpose of this choice is to maximize convenience, so circumstances where it hinders convenience, especially if it does so regularly, should be considered.

> most lines won't reach this limit, but there are cases where we need it, so we should just allow them.

This will happen no matter what, as black allows lines to exceed the limit in some circumstances where standards conflict.

> As I said before, and I repeat, comments will benefit the most from this, given that we will have more space to write them without having to split over lines

Comments won't be extended either way [[1](https://www.python.org/dev/peps/pep-0008/#id19)]. PEP-8 allows for extending lines of code but not comments or docstrings.

**Comment by leotrs** [comment]
As per @XorUnison's suggestion, I move that we have a ranked choice poll and implement the winner option.

**Comment by eulertour** [comment]
I would rather discuss and/or research the topic than choose it democratically.

**Comment by leotrs** [comment]
Sounds good then!

**Comment by XorUnison** [comment]
> I would rather discuss and/or research the topic than choose it democratically.

Right, but what exactly to discuss/research? Really the only thing I can think of in terms of researching would be how to help the people who'd be impacted by longer lines to simply not be impacted anymore, which is... easier said than done. Using IDLE and a 4K screen I can pull out well readable 291 characters per line if I really had to so I'm about as flexible as it gets.
But other people have different OSes, screens, IDEs (or lack thereof) and most importantly, preferences, with different inflexibilities, such as using terminals without even the ability to conveniently adjust fonts.
Ultimately I doubt though that "git gud" is gonna be well received with those working in a way that limits their line length, and understandably. Changing workflows tends to be annoying and unpleasant in and of itself, and given some limitations (hardware, OS) maybe even impossible.

> I'd argue that the only purpose of this choice is to maximize convenience, so circumstances where it hinders convenience, especially if it does so regularly, should be considered.

I agree.
On one hand, shorter line lengths means breaking them beyond that exact point for everyone.
On the other, longer line lengths means keeping the lines bigger than 79 characters and within the new limit intact for almost everyone, except a select few with inadequate editing conditions who will have the lines broken in place.

So the way I see it, shorter line lengths really are only more convenient for a few people, a few times, that is when those people edit that exact code.
Longer line lengths are convenient to everyone else, which I'd argue are a whole lot more people because this also includes people sifting through the code on github.
Sure, there's not that many powerusers and deep editors who'd be miffed about shorter lines, from what I see in this issue probably as much as those who'd be more miffed by longer line lengths, but with a project like this there are going to be lots and lots of people who engage with the code regularly by reading it on github. I know I do. People often have to look into the code to figure out how to implement their own niche use case, and this is one of manims selling points. And for all of those people shorter lines means more breaks like the one I mentioned twice, and thus also means more inconvenience to them.

> The fact that black occasionally allows lines to surpass the specified limit will only make it that much worse.

In regards to this, I think we should just manually enforce a hard limit, meaning when black/etc can't deal with it, we gotta deal with it on our own. And the longer the line lengths, the less often devs have to manually pick up the slack.

On one hand, shorter line lengths means less reading convenience for almost everyone which is going to be a lot more than just the core devs, and more inconvenience for devs because inevitably more manual care has to be taken with lines, particularly those using text which kinda like to ultimately fall between 80-100 characters, and black/autopep from what I gathered do not address that.
On the other, longer line lengths means less convenience just specifically to the devs with inadequate work space and inability/unwillingness to adapt, and maybe a very few select people with such restrictive screen space that they can't even view long lines properly on github.

**Comment by eulertour** [comment]
You're describing this as if there's a problem we need to solve, but the standard already has an answer to this question, namely 79. black addresses some of the problem by defaulting to 88, and it seems like we already have agreement to use that. It's fine if you don't like either of these, but it's up to you to make the case for why we should diverge. Saying things like this

> Really the only thing I can think of in terms of researching would be how to help the people who'd be impacted by longer lines to simply not be impacted anymore

> Ultimately I doubt though that "git gud" is gonna be well received with those working in a way that limits their line length

frames it as if longer lines are the default when it's actually the other way around. So when I say we should discuss or research it I just mean that we should do what we can to weigh the benefits and drawbacks, and if there's a clear reason to make the change we should.

Regarding this
> On one hand, shorter line lengths means breaking them beyond that exact point for everyone.
On the other, longer line lengths means keeping the lines bigger than 79 characters and within the new limit intact for almost everyone, except a select few with inadequate editing conditions who will have the lines broken in place.

This is an odd comparison to me since you're comparing a drawback of shorter lines to a perceived benefit of longer ones. But the benefit you describe may not actually be one, as the limit on line length also serves to prevent deep nesting and complex logic in lines that should be split up for greater clarity. Things like this

> shorter lines means more breaks like the one I mentioned twice, and thus also means more inconvenience...

make it sound like you either don't realize or don't agree with that. It's fine if you do, but you have to make the case for why having fewer line breaks is enough of a benefit to prevent encouraging line splitting for clarity.

> In regards to this, I think we should just manually enforce a hard limit, meaning when black/etc can't deal with it, we gotta deal with it on our own.

I'd highly discourage going behind our automated and standardized CI to make manual and ad-hoc changes. I can understand the sentiment behind seeing something that doesn't seem right in a standard and simply working to change it, but this is exactly what black is designed to prevent us from doing. If we make this precedent of going against the standard when we don't like it we won't get the benefit of formatting at all.

> On one hand, shorter line lengths means less reading convenience for almost everyone

I've actually heard that shorter line lengths have the opposite effect as they prevent you from moving your eyes horizontally as you read. But if this does happen to be true, the effect must be very small as projects like matplotlib, numpy, and tensorflow never saw reason to extend past 80.

**Comment by XorUnison** [comment]
> You're describing this as if there's a problem we need to solve, but the standard already has an answer to this question, namely 79.

Just the same problem you've been talking about. Convenience. Nothing less, nothing more. If it comes across like I see it as anything more than that, rest assured it's not. As I said, 88 would actually fix the worst offending lines I know.

> I'd highly discourage going behind our automated and standardized CI to make manual and ad-hoc changes.

Right, I should definitely clarify this point since that clearly came across wrong. What I mean is that they (at least autopop) can't handle splitting strings properly for instance. Unless black is different there it's really down to either us being manually aware and enforcing the standard, or the automated system inevitably failing and the PEP8 standard being violated. If it would work automated, I'd be 100% in favor of leaving it at that.

> I've actually heard that shorter line lengths have the opposite effect as they prevent you from moving your eyes horizontally as you read. But if this does happen to be true, the effect must be very small as projects like matplotlib, numpy, and tensorflow never saw reason to extend past 80.

That's... not how eye movement works? Our eyes have very small area in which we can read. If I try really hard I can read stuff I'm not directly looking at, at a distance of maybe up to 15 characters, but that's really it. You'll be moving them horizontally anyway, question is just how much.
And beyond that another question is about chunking information. While those are extremely interesting topics, do we want to delve into them that deep?

**Comment by eulertour** [comment]
> Just the same problem you've been talking about. Convenience.

What I'm saying here is that there's already an answer provided for us. If the one from the standard is bad enough that it constitutes a problem to you, it's up to you to make the case for that. Luckily, it seems like we're pretty much on the same page.

> What I mean is that they (at least autopop) can't handle splitting strings properly for instance.

My comment was referring to instances where the standards conflict. That said, I do think we should do that if black can't do it itself.

> You'll be moving them horizontally anyway, question is just how much.

Right, I should have clarified.

If that's all fine, can we settle on 88?

**Comment by leotrs** [comment]
I'm on board for 88.

**Comment by PgBiel** [comment]
I guess we can use that, and can expand if needed

**Comment by XorUnison** [comment]
My ranked order of choices would be 99>88>79, so 88 is mediocre.
But am I actually bothered enough to put in the effort to actively keep arguing for 99? I mean I definitely would if it's as opposed to 79, because that just simply leaves the code intolerably messy in some places, notably when lots of indentations, longer variable/method names and some text come into play. And it's very hard or straight up impossible to work around that while maintaining optimal readability when indentations get heavy.

But I've also checked 88 up against the code for the new geometric classes I'll PR once they're refined enough to have others look over it, and it kinda works, despite some level 6-7 indentation and long names such as `ArcBetweenPoints`. There's some cases I have to sacrifice opening delimiter alignment but there's nothing that really rubs me the wrong way. That would probably require 9-10 indentations before things really get ugly with 88, and since I don't see that happening or necessary anywhere in manim now, or in the future, I guess I'll just give a shrug to 88. Not perfect, but workable.
